src/compiler/transformers/declarations.ts

```diff
@@ -725,45 +741,197 @@ export function transformDeclarations(context: TransformationContext) {
-        let oldDiag: typeof getSymbolAccessibilityDiagnostic;
-        if (!suppressNewDiagnosticContexts) {
-            oldDiag = getSymbolAccessibilityDiagnostic;
-            getSymbolAccessibilityDiagnostic = createGetSymbolAccessibilityDiagnosticForNode(node);
+        function inferExpressionType<T extends Expression>(node: T) {
+            return resolver.createTypeOfExpression(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker) ?? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
         }
-        let typeNode;
-        switch (node.kind) {
-            case SyntaxKind.Parameter:
-            case SyntaxKind.PropertySignature:
-            case SyntaxKind.PropertyDeclaration:
-            case SyntaxKind.BindingElement:
-            case SyntaxKind.VariableDeclaration:
-                typeNode = resolver.createTypeOfDeclaration(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker, shouldAddImplicitUndefined);
-                break;
-            case SyntaxKind.FunctionDeclaration:
-            case SyntaxKind.ConstructSignature:
-            case SyntaxKind.MethodSignature:
-            case SyntaxKind.MethodDeclaration:
-            case SyntaxKind.GetAccessor:
-            case SyntaxKind.CallSignature:
-                typeNode = resolver.createReturnTypeOfSignatureDeclaration(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker);
-                break;
-            default:
-                Debug.assertNever(node);
+
+        function inferReturnTypeOfSignatureSignature<T extends SignatureDeclaration>(node: T) {
+            return resolver.createReturnTypeOfSignatureDeclaration(node, enclosingDeclaration, declarationEmitNodeBuilderFlags, symbolTracker) ?? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
         }
+        function inferAccessorType(node: GetAccessorDeclaration | SetAccessorDeclaration) {
+            if (node.kind === SyntaxKind.GetAccessor) {
+                return inferReturnTypeOfSignatureSignature(node);
+            }
+            else {
+                return factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
+            }
+        }
+        function typeFromDeclaration(node: HasInferredType | ExportAssignment) {
+            const savedErrorNameNode = errorNameNode;
+            const savedErrorFallbackNode = errorFallbackNode;
 
-        errorNameNode = undefined;
-        if (!suppressNewDiagnosticContexts) {
-            getSymbolAccessibilityDiagnostic = oldDiag!;
+            errorNameNode = node.name;
+            errorFallbackNode = node;
+            let oldDiag;
+            if (!suppressNewDiagnosticContexts) {
+                oldDiag = getSymbolAccessibilityDiagnostic;
+                getSymbolAccessibilityDiagnostic = createGetSymbolAccessibilityDiagnosticForNode(node);
+            }
+            let typeNode: TypeNode | undefined;
+            let declaredType: TypeNode | undefined;

```

Comment:
This only exists to hold what amounts to be a temporary variable, and only in two of the branches below. I would suggest the minor refactors below so we can remove this.